perceiver/data/audio/midi_processor.py

```diff
@@ -0,0 +1,275 @@
+            sustains[-1].end = ctrl.time
+    return sustains
+
+
+def _note_preprocess(susteins, notes):
+    note_stream = []
+
+    for sustain in susteins:
+        for note_idx, note in enumerate(notes):
+            if note.start < sustain.start:
+                note_stream.append(note)
+            elif note.start > sustain.end:
+                notes = notes[note_idx:]
+                sustain.transposition_notes()
+                break
+            else:
+                sustain.add_managed_note(note)
+
+    for sustain in susteins:
+        note_stream += sustain.managed_notes
+
+    note_stream.sort(key=lambda x: x.start)
+    return note_stream
+
+
+def encode_midi(midi: pretty_midi.PrettyMIDI) -> List[int]:
+    events = []
+    notes = []
+
+    for inst in midi.instruments:
+        inst_notes = inst.notes
+        # ctrl.number is the number of sustain control. If you want to know abour the number type of control,
+        # see https://www.midi.org/specifications-old/item/table-3-control-change-messages-data-bytes-2
+        ctrls = _control_preprocess([ctrl for ctrl in inst.control_changes if ctrl.number == 64])
+        if ctrls:
+            notes += _note_preprocess(ctrls, inst_notes)
+        else:
+            notes += inst_notes
+
+    dnotes = _divide_note(notes)
+
+    # print(dnotes)
+    dnotes.sort(key=lambda x: x.time)
+    # print('sorted:')
+    # print(dnotes)
+    cur_time = 0
+    cur_vel = 0
+    for snote in dnotes:
+        events += _make_time_sift_events(prev_time=cur_time, post_time=snote.time)
+        events += _snote2events(snote=snote, prev_vel=cur_vel)
+        # events += _make_time_sift_events(prev_time=cur_time, post_time=snote.time)
+
+        cur_time = snote.time
+        cur_vel = snote.velocity
+
+    return [e.to_int() for e in events]
+
+
+def decode_midi(idx_array, file_path=None) -> pretty_midi.PrettyMIDI:
+    event_sequence = [Event.from_int(idx) for idx in idx_array]
+    snote_seq = _event_seq2snote_seq(event_sequence)
+    note_seq = _merge_note(snote_seq)
+    note_seq.sort(key=lambda x: x.start)
+
+    mid = pretty_midi.PrettyMIDI()
+    # if want to change instument, see https://www.midi.org/specifications/item/gm-level-1-sound-set
+    instument = pretty_midi.Instrument(1, False, "Developed By Yang-Kichang")
+    instument.notes = note_seq
+
+    mid.instruments.append(instument)
+    if file_path is not None:
+        mid.write(file_path)
+    return mid
+
+
+def encode_midi_files(files: List[Path], num_workers: int) -> List[np.ndarray]:
+    """Encode a list of midi files using multiple cpu workers."""
+    with Pool(processes=num_workers) as pool:
+        res = list(tqdm(pool.imap(_encode_midi_file, files), total=len(files)))
+        return [r for r in res if r is not None]
+
+
+def _encode_midi_file(file: Path) -> Optional[np.ndarray]:
+    try:
+        midi_file = pretty_midi.PrettyMIDI(str(file))
+        return np.array(encode_midi(midi_file), dtype=np.int32)
```

Comment:
Why is `np.int32` used here when the data are alter saved as `np.int16`?

Reply:
good catch